ml-fundamentals/ml-experimentation.md

```diff
@@ -0,0 +1,147 @@
+- Determine how notebooks are stored and versioned (e.g. [strip output from Jupyter notebooks](https://github.com/kynan/nbstripout))
+
+#### Source control and folder structure benefits
+
+- Collaboration
+- Reproducibility
+- Code quality
+
+### Experiment tracking
+
+Experiment tracking tools allow data scientists and researchers to keep track of previous experiments for better understanding of the experimentation process and for the reproducibility of experiments or models.
+
+#### Types of experiment tracking frameworks
+
+Experiment tracking frameworks differ by the set of features they provide for collecting experiment metadata, and comparing and analysing experiments. In CSE, we mainly use [MLFlow](https://mlflow.org/) on [Databricks](https://databricks.com/product/managed-mlflow) or [Azure ML Experimentation](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-track-experiments). Note that some experiment tracking frameworks require a deployment, while others are SaaS.
+
+#### Experiment tracking outcomes
+
+1. Decide on an experiment tracking framework
+2. Ensure it is accessible to all users
+3. Document set-up on local environments
+4. Define datasets and evaluation in a way which will allow the comparison of all experiments. **Consistency across datasets and evaluation is paramount for experiment comparison**.
+5. Ensure full reproducibility by assuring that all required details are tracked (i.e. dataset names and versions, parameters, code, environment)
+
+#### Experiment tracking benefits
+
+- Model performance
+- Reproducibility
+- Collaboration
+- Code quality
+
+### Datasets and models abstractions
+
+By creating abstractions to building blocks (e.g., datasets, models, evaluators),
+we allow the easy introduction of new logic into the experimentation pipeline while keeping the agreed upon experimentation flow intact.
+
+These abstractions can be created using different mechanisms.
+For example, we can use Object Oriented Programming (OOP) solutions like abstract classes:
+
+- [An example from scikit-learn describing the creation of new estimators compatible with the API](https://scikit-learn.org/stable/developers/develop.html).
+- [An example from PyTorch on extending the abstract `Dataset` class](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html#dataset-class).
+
+#### Abstraction outcomes
+
+1. Different building blocks have defined APIs allowing them to be replaced or extended.
+2. Replacing building blocks does not break the original experimentation flow.
+3. Mock building blocks are used for unit tests
+4. APIs/mocks are shared with the engineering teams for integration with other modules.
+
+#### Abstraction benefits
+
+- Collaboration
+- Code quality
+- Reproducibility
+- Operationalization
+- Model performance
+
+### Model evaluation
```

Comment:
This section is structured differently from the others (Topic, Topic Outcomes, Topic Benefits). Why is that?

Reply:
Added benefits to fit the other topics